network/peer.go

```diff
@@ -784,3 +929,135 @@ func (p *peer) discardMyIP() {
 	}
 	p.Close()
 }
+
+// Read a single message from the connection.
+func (p *peer) readMsg() (Msg, error) {
+	pendingBuffer := wrappers.Packer{}
+	readBuffer, err := p.readFull(4)
+	if err != nil {
+		return nil, err
+	}
+	if len(readBuffer) != 4 {
+		return nil, fmt.Errorf("read buffer failed")
+	}
+
+	pendingBuffer.Bytes = append(pendingBuffer.Bytes, readBuffer...)
+
+	// lets figure out the size of the message..
+	size := pendingBuffer.UnpackInt()
+
+	// now lets read the message.
+	readBuffer, err = p.readFull(size)
+	if err != nil {
+		return nil, err
+	}
+	if uint32(len(readBuffer)) != size {
+		return nil, fmt.Errorf("read buffer failed")
+	}
+
+	pendingBuffer.Bytes = append(pendingBuffer.Bytes, readBuffer...)
+
+	// reset the offset after the size read.
+	pendingBuffer.Offset = 0
+
+	// unpack the bytes or error out...
+	msgBytes := pendingBuffer.UnpackBytes()
+	if pendingBuffer.Errored() {
+		return nil, pendingBuffer.Err
+	}
+
+	var msg Msg
+	msg, err = p.net.b.Parse(msgBytes)
+	if err != nil {
+		p.net.log.Debug("failed to parse new message from %s:\n%s\n%s",
+			p.id,
+			formatting.DumpBytes{Bytes: msgBytes},
+			err)
+		return nil, err
+	}
+
+	return msg, nil
+}
+
+// send the raw msg bytes
+func (p *peer) sendRaw(msg []byte) error {
+
+	// pack the message with a header.
+	packer := wrappers.Packer{Bytes: make([]byte, len(msg)+wrappers.IntLen)}
+	packer.PackBytes(msg)
+
+	// transmit the data..
+	msg = packer.Bytes
+	for len(msg) > 0 {
+		written, err := p.conn.Write(msg)
+		if err != nil {
+			return err
+		}
+		msg = msg[written:]
+	}
+	return nil
+}
+
+// read from the connection up to len bytes..  Don't stop till we got them.
+func (p *peer) readFull(len uint32) ([]byte, error) {
+	responseBuffer := make([]byte, 0, len)
+	for len > 0 {
+		readBuffer := make([]byte, len)
+		read, err := p.conn.Read(readBuffer)
```

Comment:
Could we switch to passing in a reference to the remaining slice of the byte array here to avoid copying the bytes an extra time?

Reply:
would go properly reference a slice?

Reply:
I believe so, a slice in go is just a reference to a portion of the underlying array. Although, I'm not sure how Read handles the case where we go past the end of the buffer. It seems like this is would be possible either way although this change could make it more likely to occur if it is an issue. I see this is the way we currently do it in the code, so perhaps this should be a separate PR to experiment with this.